src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,78 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneWheelsList(wheels);
+        this.engine = engine == null ? null : engine.clone();
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return cloneWheelsList(wheels);
+    }
+
+    public Engine getEngine() {
+        return engine == null ? null : engine.clone();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Car car = (Car) o;
+        return year == car.year && color.equals(car.color)
+                && wheels.equals(car.wheels) && engine.equals(car.engine);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(year, color, wheels, engine);
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String color) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> list = new ArrayList<>(wheels);
+        list.add(wheel);
+        return new Car(year, color, list, engine);
+    }
+
+    private List<Wheel> cloneWheelsList(List<Wheel> wheels) {
+        List<Wheel> list = new ArrayList<>();
```

Comment:
```suggestion
        List<Wheel> list = new ArrayList<>(wheels.size());
```
if you know that initial capacity will be more than 10, you can pass the capacity value in constructor. it will benefit performance